File: Answers-Ruby/040.rb
# Project Euler Problem 40, Ruby
#
# An irrational decimal fraction is created by concatenating the positive integers:
#
# 0.123456789101112131415161718192021...
#
# It can be seen that the 12th digit of the fractional part is 1.
#
# If dn represents the nth digit of the fractional part, find the value of the following expression.
#
# d1 × d10 × d100 × d1000 × d10000 × d100000 × d1000000
#
# Answer: 210

TARGET_INDICES = [1, 10, 100, 1000, 10000, 100000, 1000000]
NUM_TARGETS = 7

# We will be multiplying later, so a value of one here
finalProduct = 1
lastNumber = 0

# Use a string to assemble the sequence
assembly = ""
maximumDigitCount = TARGET_INDICES[NUM_TARGETS-1];

while (assembly.chars.count <= maximumDigitCount)
	lastNumber += 1
	
	# When you use += ruby creates a temporal object which is result of str1 + str2. 
	# Then it overrides str1 variable with reference to the new built object. 
	# On other hand << modifies existing one.
	assembly << lastNumber.to_s
end

puts << "Reached digit goal at: #{lastNumber}"

# Find the product of the numbers at the target indices
for i in 0...7
	finalProduct *= assembly[TARGET_INDICES[i]-1].to_i
end

puts "The final product is: #{finalProduct}"

File: Answers-Python/040.py
#!/usr/bin/python

# Project Euler Problem 40
#
# An irrational decimal fraction is created by concatenating the positive integers:
#
# 0.123456789101112131415161718192021...
#
# It can be seen that the 12th digit of the fractional part is 1.
#
# If dn represents the nth digit of the fractional part, find the value of the following expression.
#
# d1 x d10 x d100 x d1000 x d10000 x d100000 x d1000000
#
# Answer: 210

TARGET_INDICES = [1, 10, 100, 1000, 10000, 100000, 1000000]


# Take an integer and create a vector of the digits
def int_to_vector(input_int):
    digits = []

    while input_int > 0:
        digits.append(input_int % 10)
        input_int = int(input_int / 10)

    digits = digits[::-1]

    return digits


# We will be multiplying later, so a value of one here
final_product = 1
last_number = 0

maximum_digit_count = TARGET_INDICES[-1]
current_digit_count = 0
target_index = 0

while current_digit_count <= maximum_digit_count:
    last_number += 1
    digits = int_to_vector(last_number)

    for it in digits:
        current_digit_count += 1

        if target_index == len(TARGET_INDICES):
            break

        if current_digit_count == TARGET_INDICES[target_index]:
            target_index += 1
            final_product *= it

print ("Reached digit goal at: %s" % last_number)
print ("The final product is: %s" % final_product)
